File: generators/montithings2cpp/src/main/java/montithings/generator/steps/trafos/patterns/MultivariateAutoregressiveAnomalyDetection.java
package montithings.generator.steps.trafos.patterns;

import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;

/**
 * Combines all input data into a single autogressive anomaly detection
 */
public class MultivariateAutoregressiveAnomalyDetection extends LinearRegression implements MultivariateAnomalyDetection {
    private final int windowSize;
    private final double tolerance;
    private final List<List<Double>> pastValues;

    public MultivariateAutoregressiveAnomalyDetection(int windowSize, double tolerance) {
        this.windowSize = windowSize;
        this.tolerance = tolerance;
        this.pastValues = new ArrayList<>();
    }

    @Override
    public List<Boolean> isAnomaly(List<Double> inputs) {
        List<Boolean> res = new ArrayList<>();

        List<Double> regressionValues = this.getRegressionValues();
        this.pastValues.add(inputs);

        for (Double input : inputs) {
            Double predictedValue = this.predictValue(input, regressionValues);

            if (predictedValue == null) {
                res.add(false);
                continue;
            }

            double err = Math.abs(predictedValue - input);

            res.add(err > this.tolerance);
        }

        return res;
    }

    private List<Double> getRegressionValues() {
        List<Double> values = new ArrayList<>();

        for (int i = 0; i < this.pastValues.size(); i++) {
            values.addAll(this.getRegressionValues(i));
        }

        return values;
    }

    private List<Double> getRegressionValues(int idx) {
        List<Double> values = new ArrayList<>();

        ListIterator<Double> li = this.pastValues.get(idx).listIterator(this.pastValues.get(idx).size());

        while (li.hasPrevious() && values.size() < this.windowSize) {
            values.add(li.previous());
        }

        return values;
    }
}

File: generators/montithings2cpp/src/main/resources/rte/montithings-RTE/patterns/MultivariateAutoregressiveAnomalyDetection.cpp
// (c) https://github.com/MontiCore/monticore
#include "MultivariateAutoregressiveAnomalyDetection.h"

std::vector<bool>
MultivariateAutoregressiveAnomalyDetection::is_anomaly(std::vector<float> inputs,
                                                       std::vector<std::vector<float>> past_values)
{
  std::vector<bool> res(inputs.size());

  std::vector<float> regression_values = this->get_regression_values(past_values);

  for (int i = 0; i < inputs.size(); i++)
  {
    float predicted_value = this->lr->predict_value(inputs[i], regression_values);

    float err = std::abs(predicted_value - inputs[i]);

    res[i] = err > this->tolerance;
  }

  return res;
}

std::vector<float>
MultivariateAutoregressiveAnomalyDetection::get_regression_values(
    std::vector<std::vector<float>> past_values)
{
  int values_len = std::min({this->window_size, static_cast<int>(past_values.size())});

  std::vector<float> values;

  int i = past_values.size();
  int count = 0;

  while (count < values_len)
  {
    count++;
    i--;

    for (int j = 0; j < past_values[i].size(); j++)
    {
      values.push_back(past_values[i][j]);
    }
  }

  return values;
}